src/webgpu/shader/execution/expression/call/builtin/derivatives.ts

```diff
@@ -0,0 +1,218 @@
+ * @param vectorize if defined, the vector width to use (2, 3, or 4)
+ */
+export function runDerivativeTest(
+  t: GPUTest,
+  cases: Case[],
+  builtin: string,
+  non_uniform_discard: boolean,
+  vectorize?: number
+) {
+  // If the 'vectorize' config option was provided, pack the cases into vectors.
+  let type: Type = Type.f32;
+  if (vectorize !== undefined) {
+    const packed = packScalarsToVector([type, type], type, cases, vectorize);
+    cases = packed.cases;
+    type = packed.resultType;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  // The two input values for a given case are distributed to two different invocations in a quad.
+  // We will populate a storage buffer with these input values laid out sequentially:
+  // [ case_0_input_1, case_0_input_0, case_1_input_1, case_1_input_0, ...]
+  //
+  // The render pipeline will be launched several times over a viewport size of (2, 2). Each draw
+  // call will execute a single quad (four fragment invocation), which will exercise two test cases.
+  // Each of these draw calls will use a different instance index, which is forwarded to the
+  // fragment shader. Each invocation will determine its index into the storage buffer using its
+  // fragment position and the instance index for that draw call.
+  //
+  // Consider two draw calls that test 4 cases (c_0, c_1, c_2, c_3).
+  //
+  // For derivatives along the 'x' direction, the mapping from fragment position to case input is:
+  // Quad 0: | c_0_i_1 | c_0_i_0 |     Quad 1: | c_2_i_1 | c_2_i_0 |
+  //         | c_1_i_1 | c_1_i_0 |             | c_3_i_1 | c_3_i_0 |
+  //
+  // For derivatives along the 'y' direction, the mapping from fragment position to case input is:
+  // Quad 0: | c_0_i_1 | c_1_i_1 |     Quad 1: | c_2_i_1 | c_3_i_1 |
+  //         | c_0_i_0 | c_1_i_0 |             | c_2_i_0 | c_3_i_0 |
+  //
+  ////////////////////////////////////////////////////////////////
+
+  // Determine the direction of the derivative ('x' or 'y') from the builtin name.
+  const dir = builtin[3];
+
+  // Determine the WGSL type to use in the shader, and the stride in bytes between values.
+  let valueStride = 4;
+  let wgslType = 'f32';
+  if (vectorize) {
+    wgslType = `vec${vectorize}f`;
+    valueStride = vectorize * 4;
+    if (vectorize === 3) {
+      valueStride = 16;
+    }
+  }
+
+  // Define a vertex shader that draws a quad over the full viewport, and a fragment shader that
+  // calls the derivative builtin with a value loaded from that fragment's index into the storage
+  // buffer (determined using the quad index and fragment position, as described above).
+  const code = `
+struct CaseInfo {
+  @builtin(position) position: vec4f,
+  @location(0) @interpolate(flat) quad_idx: u32,
+}
+
+@vertex
+fn vert(@builtin(vertex_index) vertex_idx: u32,
+        @builtin(instance_index) instance_idx: u32) -> CaseInfo {
+  const kVertices = array(
+    vec2f(-1, -1),
```

Comment:
For the sake of reducing potential issues along the seam between the two triangles, perhaps it would be better to have a single triangle that covers the entire viewport. Might also be a good thing to parameterize.

Reply:
Switched to a single triangle.